### PR TITLE
reaper: mark audio backend libs as optional, add missing plugin libs

### PR DESCRIPTION
**Pull request by ilian**
###### Motivation for this change
The JACK and PulseAudio audio backends are optional. This change allows users to enable or disable support for them.

This change also adds optional libraries that are needed to use the included plugins:
- LAME for mp3 support
- VLC for video decoding
- FFmpeg for video encoding and decoding

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by SuperSandro2000** [comment]
> The JACK and PulseAudio audio backends are optional. This change allows users to enable or disable support for them.

Is there a compelling reason to disable them? Options which serve no real benefit should be avoided.

File: pkgs/applications/audio/reaper/default.nix [hunk 1]
```diff
@@ -1,10 +1,43 @@
...
+, makeWrapper
+
+, alsaLib
+, gtk3
+
+, jackSupport ? true, libjack2 ? null
+, pulseaudioSupport ? true, libpulseaudio ? null
+, mp3Support ? true, lame ? null
+, ffmpegSupport ? true, ffmpeg ? null
+, vlcSupport ? true, vlc ? null
```
Comment (SuperSandro2000) [inline pkgs/applications/audio/reaper/default.nix:13]:
```suggestion
, jackSupport ? true, libjack2
, pulseaudioSupport ? true, libpulseaudio
, mp3Support ? true, lame
, ffmpegSupport ? true, ffmpeg
, vlcSupport ? true, vlc
```

File: pkgs/applications/audio/reaper/default.nix [hunk 2]
```diff
@@ -1,10 +1,43 @@
...
 }:
 
-stdenv.mkDerivation rec {
+assert jackSupport -> libjack2 != null;
+assert pulseaudioSupport -> libpulseaudio != null;
+assert mp3Support -> lame != null;
+assert ffmpegSupport -> ffmpeg != null;
+assert vlcSupport -> vlc != null;
```
Comment (SuperSandro2000) [inline pkgs/applications/audio/reaper/default.nix:20]:
```suggestion
```

Not required.

File: pkgs/applications/audio/reaper/default.nix [hunk 3]
```diff
@@ -1,10 +1,43 @@
...
+
+let
+  inherit (lib) optional;
+
```
Comment (SuperSandro2000) [inline pkgs/applications/audio/reaper/default.nix:24]:
```suggestion
```

File: pkgs/applications/audio/reaper/default.nix [hunk 4]
```diff
@@ -1,10 +1,43 @@
...
+  # Dynamic loading of plugin dependencies does not adhere to rpath of
+  # reaper executable that gets modified with runtimeDependencies.
+  # Patching each plugin with DT_NEEDED is cumbersome and requires
+  # hardcoding of API versions of each dependency.
+  # Setting the rpath of the plugin shared object files does not
+  # seem to have an effect for some plugins.
+  # We opt for wrapping the executable with LD_LIBRARY_PATH prefix.
+  pluginDeps = optional mp3Support lame
+    ++ optional ffmpegSupport ffmpeg
+    ++ optional vlcSupport vlc;
```
Comment (SuperSandro2000) [inline pkgs/applications/audio/reaper/default.nix:34]:
```suggestion
  pluginDeps = lib.optional mp3Support lame
    ++ lib.optional ffmpegSupport ffmpeg
    ++ lib.optional vlcSupport vlc;
```

File: pkgs/applications/audio/reaper/default.nix [hunk 5]
```diff
@@ -1,10 +1,43 @@
...
+  wrapCommand = if builtins.length pluginDeps == 0 then "" else ''
```
Comment (SuperSandro2000) [inline pkgs/applications/audio/reaper/default.nix:35]:
```suggestion
  wrapCommand = lib.optionalString (builtins.length pluginDeps != 0) ''
```

File: pkgs/applications/audio/reaper/default.nix [hunk 6]
```diff
@@ -17,22 +50,15 @@ stdenv.mkDerivation rec {
...
+    gtk3
   ];
 
   runtimeDependencies = [
-    gtk3
-  ];
+    gtk3 # libSwell needs libgdk-3.so.0
+  ]
+  ++ optional jackSupport libjack2
+  ++ optional pulseaudioSupport libpulseaudio;
```
Comment (SuperSandro2000) [inline pkgs/applications/audio/reaper/default.nix:61]:
```suggestion
  ++ lib.optional jackSupport libjack2
  ++ lib.optional pulseaudioSupport libpulseaudio;
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

**Comment by ilian** [comment]
> > The JACK and PulseAudio audio backends are optional. This change allows users to enable or disable support for them.
> 
> Is there a compelling reason to disable them? Options which serve no real benefit should be avoided.

Adding such an option might be useful to avoid having to include PulseAudio and JACK on a system where reaper exclusively manages the audio device through ALSA.

**Comment by SuperSandro2000** [comment]
> Adding such an option might be useful to avoid having to include PulseAudio and JACK on a system where reaper exclusively manages the audio device through ALSA.

This would mean we would only need an option like disablePulseAudio. I want to avoid adding options for things no one uses or tests.

**Comment by ilian** [comment]
> > Adding such an option might be useful to avoid having to include PulseAudio and JACK on a system where reaper exclusively manages the audio device through ALSA.
> 
> This would mean we would only need an option like disablePulseAudio. I want to avoid adding options for things no one uses or tests.

I don't see why we should add an option for only one of the two backends that are marked as optional in the `readme-linux.txt` inside the binary tarball.
I think the `jackSupport` and `pulseaudioSupport` options make most sense to me instead of inverting the option since it is consistent with many existing packages that use jack. Examples include:
- https://github.com/NixOS/nixpkgs/blob/d8d06d1cec862335f303d823dea6753488b8c944/pkgs/applications/audio/fmit/default.nix#L3
- https://github.com/NixOS/nixpkgs/blob/e2fcf81d2a312f0bf83a7939988db2e63f79cf9f/pkgs/applications/audio/cmus/default.nix#L7-L10
- https://github.com/NixOS/nixpkgs/blob/e2fcf81d2a312f0bf83a7939988db2e63f79cf9f/pkgs/applications/audio/moc/default.nix#L5-L6

**Comment by SuperSandro2000** [comment]
> I don't see why we should add an option for only one of the two backends that are marked as optional in the readme-linux.txt inside the binary tarball.


Because we do not follow those upstream recommendations all the time. Usually nixpkgs packages are build with all available options if they work together and there is no other compelling reason to disable them. Incase of qemu there is a full packages with almost all options and a slim one with only the most necessary options for tests IIRC.

File: pkgs/applications/audio/reaper/default.nix [hunk 7]
```diff
@@ -1,10 +1,35 @@
...
+, makeWrapper
+
+, alsaLib
+, gtk3
+
+, jackSupport ? true, libjack2
+, pulseaudioSupport ? config.pulseaudio or true, libpulseaudio
+, mp3Support ? true, lame
+, ffmpegSupport ? true, ffmpeg
+, vlcSupport ? true, vlc
```
Comment (SuperSandro2000) [inline pkgs/applications/audio/reaper/default.nix:13]: Do we really need options to disable mp3, ffmpeg and vlc support? Such options are normally not tested regularly and easily bit rot over time.

File: pkgs/applications/audio/reaper/default.nix [hunk 8]
```diff
@@ -1,10 +1,37 @@
...
+  # hardcoding of API versions of each dependency.
+  # Setting the rpath of the plugin shared object files does not
+  # seem to have an effect for some plugins.
+  # We opt for wrapping the executable with LD_LIBRARY_PATH prefix.
+  pluginDeps = [
+    lame
+    ffmpeg
+    vlc
+  ];
+  wrapCommand = lib.optionalString (builtins.length pluginDeps != 0) ''
```
Comment (SuperSandro2000) [inline pkgs/applications/audio/reaper/default.nix:29]: I think we inline the wrapCommand. It should make it easier for other people to follow the build commands.

File: pkgs/applications/audio/reaper/default.nix [hunk 9]
```diff
@@ -1,10 +1,37 @@
...
+    wrapProgram $out/opt/REAPER/reaper \
+      --prefix LD_LIBRARY_PATH : "${lib.makeLibraryPath pluginDeps}"
+  '';
+
```
Comment (SuperSandro2000) [inline pkgs/applications/audio/reaper/default.nix:33]:
```suggestion
```

File: pkgs/applications/audio/reaper/default.nix [hunk 10]
```diff
@@ -42,8 +62,7 @@ stdenv.mkDerivation rec {
       --integrate-user-desktop
     rm $out/opt/REAPER/uninstall-reaper.sh
 
-    wrapProgram $out/opt/REAPER/reaper \
-      --prefix LD_LIBRARY_PATH : "${lib.makeLibraryPath [ libpulseaudio libjack2 ]}"
+    ${wrapCommand}
```
Comment (SuperSandro2000) [inline pkgs/applications/audio/reaper/default.nix:65]:
```suggestion
  # Dynamic loading of plugin dependencies does not adhere to rpath of
  # reaper executable that gets modified with runtimeDependencies.
  # Patching each plugin with DT_NEEDED is cumbersome and requires
  # hardcoding of API versions of each dependency.
  # Setting the rpath of the plugin shared object files does not
  # seem to have an effect for some plugins.
  # We opt for wrapping the executable with LD_LIBRARY_PATH prefix.
    wrapProgram $out/opt/REAPER/reaper \
      --prefix LD_LIBRARY_PATH : "${lib.makeLibraryPath [ ffmpeg lame vlc ]}"
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: pkgs/applications/audio/reaper/default.nix (hunk 8 above)
Comment (ilian) [reply pkgs/applications/audio/reaper/default.nix:29]: Makes sense, thanks!

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review which is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 110582` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
1 package built:
</summary>
<ul>
<li>reaper</li>
</ul>
</details>

**Pull request merged**